File: neko/getoneko.py
#!/usr/bin/python3

import sys
import math

R = 480
OX = 960
OY = 540

def getpos(generation): 
    rad1 = math.radians((generation % 90) * 4)
    rad2 = math.radians((generation % 10) * 36)
    r = (math.cos(rad2) * 20) + R
    x = math.cos(rad1) * r
    y = math.sin(rad1) * r
    return (x, y)

if len(sys.argv) > 1: 
    now = int(sys.argv[1])
    pos_now = getpos(now)
    x_now = int(pos_now[0])
    y_now = int(pos_now[1])

    # determine which bitmap to use by comparing the 
    # current position with the next one which comes 
    # on the +1 generation. 
    pos_next = getpos(now + 1)
    x_next = int(pos_next[0])
    y_next = int(pos_next[1])
    significant = 8
    prefix = "kaki"
    west = False
    east = False
    north = False
    south = False

    if (x_now - x_next > significant): 
        west = True
    if (x_next - x_now > significant): 
        east = True
    if (y_now - y_next > significant): 
        north = True
    if (y_next - y_now > significant): 
        south = True

    if (north and east): 
        prefix = "upright"
    if (south and east): 
        prefix = "dwright"
    if (south and west): 
        prefix = "dwleft"
    if (north and west): 
        prefix = "upleft"

    if (north and not (west or east)): 
        prefix = "up"
    if (east and not (north or south)): 
        prefix = "right"
    if (south and not (west or east)): 
        prefix = "down"
    if (west and not (north or south)): 
        prefix = "left"
    if (not (north or east or south or west)): 
        prefix = "kaki"

    movement = int((now % 2) + 1)
    filename = prefix + str(movement) + ".png"

    # print the result in CSV format. 
    print(str(x_now + OX) + "," + str(y_now + OY) + "," + filename)
else: 
    print("FATAL: no arguments specified.")

File: cat/js/getoneko.js
const R = 480;
const OX = 960;
const OY = 540;

function radians(deg) {
    return deg * Math.PI / 180;
}

function getpos(t) {
    const rad1 = radians((t % 90) * 4);
    const rad2 = radians((t % 10) * 36);
    const r = (Math.cos(rad2) * 20) + R;
    const x = Math.cos(rad1) * r;
    const y = Math.sin(rad1) * r;

    return [x, y];
}

function getoneko(now) {
    var ret = Object();
    const pos_now = getpos(now);
    const x_now = Math.floor(pos_now[0]);
    const y_now = Math.floor(pos_now[1]);
    const pos_next = getpos(now + 1);
    const x_next = Math.floor(pos_next[0]);
    const y_next = Math.floor(pos_next[1]);

    ret.x = OX + x_now;
    ret.y = OY + y_now;
    ret.filnm = "";

    /*
        determine which bitmap to use by comparing the 
        current position with the next one which comes 
        on the +1 generation. 
    */
    const significant = 8;
    var prefix = 10;
    var west = false;
    var east = false;
    var north = false;
    var south = false;
    var movement = 0;

    if (x_now - x_next > significant) { 
        west = true;
    }
    if (x_next - x_now > significant) { 
        east = true;
    }
    if (y_now - y_next > significant) { 
        north = true;
    }
    if (y_next - y_now > significant) { 
        south = true;
    }

    if (north && east) { 
        prefix = 28;
    }
    if (south && east) {
        prefix = 7;
    }
    if (south && west) {
        prefix = 5;
    }
    if (north && west) {
        prefix = 26;
    }

    if (north && !(west || east)) {
        prefix = 24;
    }
    if (east && !(north || south)) {
        prefix = 18;
    }
    if (south && !(west || east)) {
        prefix = 1;
    }
    if (west && !(north || south)) {
        prefix = 12;
    }
    if (!(north || east || south || west)) {
        prefix = 10;
    }

    movement = Math.floor(now % 2);
    ret.imgid = prefix + movement;

    return ret;
}

